src/kakarot/instructions/stop_and_math_operations.cairo

```diff
@@ -0,0 +1,565 @@
+        range_check_ptr,
+        bitwise_ptr: BitwiseBuiltin*,
+    }(ctx: model.ExecutionContext*) -> model.ExecutionContext* {
+        // return_data stored the return_data for the last executed sub context
+        // see CALLs opcodes. When we run the STOP opcode, we stop the current
+        // execution context with *no* return data (unlike RETURN and REVERT).
+        // hence we just clear the return_data and stop.
+        let (return_data: felt*) = alloc();
+        let ctx = ExecutionContext.stop(ctx, 0, return_data, FALSE);
+        return ctx;
+    }
+
+    func exec_math_operation{
+        syscall_ptr: felt*,
+        pedersen_ptr: HashBuiltin*,
+        range_check_ptr,
+        bitwise_ptr: BitwiseBuiltin*,
+    }(ctx: model.ExecutionContext*) -> model.ExecutionContext* {
+        alloc_locals;
+
+        local opcode: model.Opcode*;
+
+        // See evm.cairo, pc is increased before entering the opcode
+        let opcode_number = [ctx.call_context.bytecode + ctx.program_counter - 1];
+
+        // To cast the codeoffset opcodes_label to a model.Opcode*, we need to use it to offset
+        // the current pc. We get the pc from the `get_fp_and_pc` util and assign a codeoffset (pc_label) to it.
+        // In short, this boilds down to: opcode = pc + offset - pc = offset
+        let (_, pc) = get_fp_and_pc();
+
+        pc_label:
+        assert opcode = cast(
+            pc + (opcodes_label - pc_label) + opcode_number * model.Opcode.SIZE, model.Opcode*
+        );
+
+        let stack_underflow = is_le(ctx.stack.size, opcode.stack_input - 1);
+        if (stack_underflow != 0) {
+            let (revert_reason_len, revert_reason) = Errors.stackUnderflow();
+            let ctx = ExecutionContext.stop(ctx, revert_reason_len, revert_reason, TRUE);
+            return ctx;
+        }
+        let out_of_gas = is_le(ctx.call_context.gas_limit, ctx.gas_used + opcode.gas - 1);
+        if (out_of_gas != 0) {
+            let (revert_reason_len, revert_reason) = Errors.outOfGas();
+            let ctx = ExecutionContext.stop(ctx, revert_reason_len, revert_reason, TRUE);
+            return ctx;
+        }
+
+        let (local stack, popped) = Stack.pop_n(ctx.stack, opcode.stack_input);
+
+        // offset is 1 (new line) + 2 (jmp + label) per opcode
+        // opcode is offset from by 0x1 (index of the first opcode)
+        tempvar offset = 2 * (opcode_number - 0x01) + 1;
+
+        tempvar range_check_ptr = range_check_ptr;
+        tempvar popped = popped;
+
+        jmp rel offset;
+
+        jmp ADD;  // 0x1
+        jmp MUL;  // 0x2
+        jmp SUB;  // 0x3
+        jmp DIV;  // 0x4
+        jmp SDIV;  // 0x5
+        jmp MOD;  // 0x6
+        jmp SMOD;  // 0x7
+        jmp ADDMOD;  // 0x8
+        jmp MULMOD;  // 0x9
+        jmp EXP;  // 0xa
+        jmp SIGNEXTEND;  // 0xb
+        jmp INVALID;  // 0xc
+        jmp INVALID;  // 0xd
+        jmp INVALID;  // 0xe
+        jmp INVALID;  // 0xf
```

Comment:
what happens if we jump to `INVALID`? Does this cause a cairo vm error?

Reply:
uhm rn INVALID is like LT (the one right after). This function is not meant to be called with any opcode number though (see evm.cairo) but I needed to have a label. If/when the main evm.cairo is updated to something like this kind of jump table, it'd be updated.